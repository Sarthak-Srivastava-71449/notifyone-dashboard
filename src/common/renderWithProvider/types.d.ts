/*
    Auth User Types
*/

interface IUser {
  loading: boolean;
  error: string;
  success: boolean | number;
  name: string;
  tokens: {
    accessToken: string;
  };
  user_info: {
    email_id: string;
    contact_number: string;
    email_verified: boolean;
  };
  roles: IRoles;
}

type UserAction = {
  type?: string;
  payload?: IUser;
};

type UserActionDispatchType = (args: UserAction) => UserAction | any;

interface AxiosGoogleResponse {
  auth_token: string;
  auth_token_labs: string;
  token: string;
  username: string;
}

interface UserInfoResponse {
  roles: IRoles;
  user_info: {
    email_id: string;
    contact_number: string;
    email_verified: boolean;
  };
}
