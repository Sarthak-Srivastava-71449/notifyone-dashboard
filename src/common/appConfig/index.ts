const AppConfig: any = process.env.communication;

export default AppConfig;
