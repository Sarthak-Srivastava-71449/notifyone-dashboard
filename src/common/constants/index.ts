export const COMMUNICATION_TYPE = {
  All: 'All',
  SMS: 'SMS',
  Whatsapp: 'Whatsapp',
  Transactional_Push_notification: 'Transactional Push Notification',
  Email: 'Email',
};

export const MAX_INPUT_LIMIT = 30;

export const AUTH_ROLES = 'LARA_ADMIN';

export const EVENT_FETCH_LIMIT = 50;

// Navigation
export const COMS_TEMPLATE_NAVIGATE = '/communication/template';
