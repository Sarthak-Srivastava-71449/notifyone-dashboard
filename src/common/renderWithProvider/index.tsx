export { default as withProvider } from './renderWithProvider';
export { store, withLocationProvider } from './renderWithProvider';
export {
  mockEmailEventDetails,
  mockPushNotificationEventDetails,
  mockSmsEventDetails,
  mockWhatsappEventDetails,
} from './constants';
