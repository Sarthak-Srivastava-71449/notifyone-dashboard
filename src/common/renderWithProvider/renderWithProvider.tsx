import React from 'react';
import thunk from 'redux-thunk';
import {
  applyMiddleware,
  combineReducers,
  legacy_createStore as createStore,
} from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
import { Provider } from 'react-redux';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import { IntlProvider } from 'react-intl';
import { ToastContainer } from 'react-toastify';

import messages from '../i18n/messages';
import googleReducer from 'src/platform/reducers/auth.reducer';

import emailEventsReducer from 'src/store/reducers/emailEvents.reducer';
import pushNotificationEventsReducer from 'src/store/reducers/pushNotificationEvents.reducer';
import smsEventsReducer from 'src/store/reducers/smsEvents.reducer';
import whatsAppEventsReducer from 'src/store/reducers/whatsappEvents.reducer';

import Dashboard from 'src/view/components/Dashboard/Dashboard';
import CommunicationList from 'src/view/components/CommunicationList/CommunicationList';
import Communication from 'src/view/components/Communication/Communication';

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

jest.mock('antd', () => {
  const antd = jest.requireActual('antd');

  return {
    ...antd,
  };
});

jest.mock('common/src/components/Spinner', () => ({
  Spinner: () => {
    return <></>;
  },
}));

let store: any;

const createTestStore = () => {
  const store = createStore(
    combineReducers({
      emailEvents: emailEventsReducer,
      pushNotificationEvents: pushNotificationEventsReducer,
      smsEvents: smsEventsReducer,
      whatsAppEvents: whatsAppEventsReducer,
      user: googleReducer,
    }),
    composeWithDevTools(applyMiddleware(thunk))
  );
  return store;
};

const withProvider = (component: any) => {
  store = createTestStore();
  store.dispatch = jest.fn();
  const locale = 'en-us';

  return (
    <Provider store={store}>
      <IntlProvider locale={locale} messages={messages[locale]}>
        <BrowserRouter>{component}</BrowserRouter>
      </IntlProvider>
    </Provider>
  );
};

const withLocationProvider = (to: string, state: any = {}) => {
  store = createTestStore();
  store.dispatch = jest.fn();
  const locale = 'en-us';

  window.history.pushState(state, '', to);

  return (
    <Provider store={store}>
      <IntlProvider locale={locale} messages={messages[locale]}>
        <BrowserRouter>
          <ToastContainer />
          <Routes>
            <Route path="communication" element={<Dashboard />}>
              <Route path="templates" element={<CommunicationList />} />
              <Route path="template" element={<Communication />} />
            </Route>
          </Routes>
        </BrowserRouter>
      </IntlProvider>
    </Provider>
  );
};

export { store, withLocationProvider };
export default withProvider;
