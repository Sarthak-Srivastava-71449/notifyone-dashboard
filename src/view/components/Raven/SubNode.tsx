import { Col, Collapse, Input, Modal, Popconfirm, Row, Select } from 'antd';
import React, { useEffect, useRef, useState } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { Button, ButtonType } from '../Button';
import authPermissionHandler from 'src/common/authPermission/authPermissions';
import appNames from 'src/common/constants/appNames';
import rightConstants from 'src/common/constants/rightConstants';
import { RootState } from 'src/store';
import { updateNewNodeAction } from 'src/store/actions/ravenActionNodeEvents';
import { sendNodeLink } from 'src/store/actions/ravenNodeEvents';
import {
  assignRavenRootNodeDetails,
  createChildNodeEvent,
  getLinkedNodeEvents,
  onChangeNodeRank,
  setCurrentNodekey,
  updateRankEvent,
} from 'src/store/actions/ravenRootNodeEvents';
import { getAccessToken } from 'src/store/selectors/accessToken';
import {
  getNewNodeOptions,
  getNodeActions,
} from 'src/store/selectors/createNode';
import { getInactiveNodes, getUserRoles } from 'src/store/selectors/subNode';
import './Raven.css';
import { toast } from 'react-toastify';

const { Option } = Select;

interface SubNodeProps {
  nodeSubQuestion: any;
  rootNode?: boolean;
}

const SubNode: React.FC<SubNodeProps> = ({
  nodeSubQuestion,
  rootNode,
}: SubNodeProps) => {
  const [collapseOpen, setCollapseOpen] = useState(false);
  const [selectedNodeActionId, setSelectedNodeActionId] = useState<
    number | null
  >(null);
  const [showModal, setShowModal] = useState(false);
  const [showModalChangeAction, setShowModalChangeAction] = useState(false);
  const rankInput = useRef<any>(null);

  /*
  Redux Selectors for fetching data
 */
  const userRoles = useSelector((state: RootState) => getUserRoles(state));
  const inactiveNodes = useSelector((state: RootState) =>
    getInactiveNodes(state)
  );
  const accessToken = useSelector((state: RootState) => getAccessToken(state));
  const newNodeOptions = useSelector((state: RootState) =>
    getNewNodeOptions(state)
  );
  const nodeActions = useSelector((state: RootState) => getNodeActions(state));
  const currentNodeKey = useSelector(
    (state: RootState) => state?.ravenRootEventsReducer?.currentNodeKey
  );
  const updatedRankArr = useSelector(
    (state: RootState) => state?.ravenRootEventsReducer?.updatedRank
  );

  useEffect(() => {
    if (rankInput?.current !== null && updatedRankArr?.length === 0) {
      rankInput.current.input.value = '';
    }
  }, [updatedRankArr, rankInput]);

  const dispatch = useDispatch();
  const intl = useIntl();
  const { Panel } = Collapse;

  const updateNodeLink = (nodeName: string, activationStatus: string) => {
    const payload = {
      active: activationStatus,
      node_name: nodeName,
    };
    dispatch(sendNodeLink(accessToken, payload));
  };

  const onChangeAddNode = (event: any) => {
    // If you don't want click extra trigger collapse, you can prevent this:
    event.stopPropagation();
    setShowModal(true);
  };

  const handleChange = (nodeActionValueId: number) => {
    setSelectedNodeActionId(nodeActionValueId);
  };

  const handleChangeDone = (nodeName: string) => {
    const payload = {
      child_id: selectedNodeActionId?.toString(),
      nodeAction: selectedNodeActionId?.toString(),
      parent_name: nodeName,
      rank: null,
    };
    setShowModal(false);
    dispatch(createChildNodeEvent(accessToken, payload));
  };

  const closeModal = () => {
    setShowModal(false);
  };

  const onChangeAction = () => {
    setShowModalChangeAction(true);
  };

  const handleChangeAction = (nodeActionValueId: number) => {
    setSelectedNodeActionId(nodeActionValueId);
  };

  const handleChangeActionDone = (nodeName: string) => {
    setShowModalChangeAction(false);

    const payload = {
      child_id: selectedNodeActionId?.toString(),
      nodeAction: selectedNodeActionId?.toString(),
      parent_name: nodeName,
    };
    dispatch(updateNewNodeAction(accessToken, payload));
  };

  const closeModalChangeAction = () => {
    setShowModalChangeAction(false);
  };
  const sideLink = (nodeSubQuestion: any) => (
    <div
      className="flex link items-center"
      onClick={(event) => {
        // If you don't want click extra trigger collapse, you can prevent this:
        event.stopPropagation();
      }}
    >
      <div className="inner-item">
        {!rootNode && (
          <>
            <div className="px-1 font-semibold text-sm">
              {/* This is a button which is used as a toggle */}
              {inactiveNodes &&
              inactiveNodes?.length > 0 &&
              inactiveNodes.includes(nodeSubQuestion?.id) ? (
                <a
                  className="condition-btn condition-btn-true px-1 py-1 link"
                  onClick={() => updateNodeLink(nodeSubQuestion.id, 'true')}
                >
                  <FormattedMessage id="true" />
                </a>
              ) : (
                <a
                  className="condition-btn px-1 py-1 link"
                  onClick={() => updateNodeLink(nodeSubQuestion.id, 'false')}
                >
                  <FormattedMessage id="false" />
                </a>
              )}
            </div>
            <div className="ml-3">
              <Input
                type="number"
                className={
                  'custom-input input-number focus:ring-blue-500 focus:outline-none appearance-none w-full text-sm leading-6  rounded-md py-1 px-1.5  shadow-sm'
                }
                placeholder={intl?.formatMessage({ id: 'rank' })}
                data-testid="rank-number"
                ref={rankInput}
                onBlur={(e) => {
                  const v = e.target.value;
                  const parentId = nodeSubQuestion.id.slice(
                    0,
                    nodeSubQuestion.id.lastIndexOf('.') // get the parent id -> 161.138.1 gives 161.138
                  );
                  if (currentNodeKey !== null && parentId === currentNodeKey) {
                    if (v) {
                      dispatch(
                        onChangeNodeRank(v.toString(), nodeSubQuestion.id)
                      );
                    }
                  } else {
                    toast.info('Reopen the required sub panel again!');
                  }
                }}
              />
            </div>
            <div className="ml-3">
              {!nodeSubQuestion?.sub_questions && (
                <a className="link ml-1.5 px-1" onClick={onChangeAction}>
                  <FormattedMessage id="change_action" />
                </a>
              )}
            </div>
          </>
        )}
        {authPermissionHandler(
          userRoles,
          appNames?.HELP,
          rightConstants?.UPDATE
        ) && (
          <div className="edit-node-btn">
            {/* This is for Edit Action */}
            {!nodeSubQuestion.sub_questions ? (
              <Link
                rel="noopener noreferrer"
                replace={true}
                to={`/communication/raven/edit/action/${nodeSubQuestion.id}`}
              >
                <Button
                  className="px-1"
                  onClick={() => {
                    dispatch(assignRavenRootNodeDetails(nodeSubQuestion));
                    dispatch(
                      getLinkedNodeEvents(
                        accessToken,
                        nodeSubQuestion.node_action
                      )
                    );
                  }}
                >
                  <FormattedMessage id="edit_action" />
                </Button>
              </Link>
            ) : (
              // This is for Edit Node
              <Link
                rel="noopener noreferrer"
                replace={true}
                className="edit-btn"
                to={`/communication/raven/edit/${nodeSubQuestion.id}`}
              >
                <Button
                  className="px-1"
                  onClick={() => {
                    dispatch(assignRavenRootNodeDetails(nodeSubQuestion));
                  }}
                >
                  <span className="parent-edit-btn">
                    <FormattedMessage id="edit" />
                  </span>
                  <span className="child-edit-btn">
                    <FormattedMessage id="edit_node" />
                  </span>
                </Button>
              </Link>
            )}
          </div>
        )}
      </div>
    </div>
  );

  const RootNode = (subQuestionLength: number | null) => (
    <div className="node-link flex">
      {authPermissionHandler(
        userRoles,
        appNames.HELP,
        rightConstants.CREATE
      ) && (
        <Button
          className="text-sm mr-1.5 link px-1.5"
          onClick={(e) => onChangeAddNode(e)}
        >
          <FormattedMessage id="add_node" />
        </Button>
      )}
      <Button
        className="text-sm link px-1.5"
        disabled={
          subQuestionLength === null ||
          updatedRankArr.length !== subQuestionLength
        }
        onClick={() => dispatch(updateRankEvent(accessToken, updatedRankArr))}
      >
        <FormattedMessage id="update_rank" />
      </Button>
    </div>
  );

  let subQuestionFlag = false;
  if (
    nodeSubQuestion?.sub_questions &&
    nodeSubQuestion?.sub_questions?.length > 0
  ) {
    subQuestionFlag = false;
  } else {
    subQuestionFlag = true;
  }
  return (
    <div>
      <Collapse
        accordion
        onChange={(key: any) => {
          const v = key !== undefined ? key : null; // if key exists then set key else set to null
          dispatch(setCurrentNodekey(v));
          setCollapseOpen(!collapseOpen);
        }}
      >
        {!collapseOpen ? (
          <Panel
            header={nodeSubQuestion?.id + ' ' + nodeSubQuestion?.name}
            key={nodeSubQuestion?.id}
            extra={sideLink(nodeSubQuestion)}
            disabled={subQuestionFlag}
            className={
              inactiveNodes?.includes(nodeSubQuestion?.id) && 'bg-zinc-100'
            }
          >
            {nodeSubQuestion?.sub_questions &&
              nodeSubQuestion.sub_questions?.length > 0 &&
              nodeSubQuestion.sub_questions.map((subNodeQuestions: any) => (
                <SubNode
                  key={subNodeQuestions?.id}
                  nodeSubQuestion={subNodeQuestions}
                />
              ))}
          </Panel>
        ) : (
          <Panel
            header={nodeSubQuestion.name}
            key={nodeSubQuestion.id}
            extra={RootNode(nodeSubQuestion?.sub_questions?.length || null)}
            disabled={subQuestionFlag}
          >
            {nodeSubQuestion.sub_questions &&
              nodeSubQuestion.sub_questions?.length > 0 &&
              nodeSubQuestion.sub_questions.map((nodeSubQuestion: any) => (
                <SubNode
                  key={nodeSubQuestion.id}
                  nodeSubQuestion={nodeSubQuestion}
                />
              ))}
          </Panel>
        )}
      </Collapse>
      <Modal
        title={intl?.formatMessage({ id: 'select_node' })}
        centered
        className="subnode-wrapper"
        visible={showModal}
        onCancel={closeModal}
        width={700}
        destroyOnClose
        footer={
          <Row className="flex justify-end">
            <Col>
              <Popconfirm
                className="popconfrim"
                title="Are you sure？"
                onConfirm={() => handleChangeDone(nodeSubQuestion.id)}
                okText="Yes"
                cancelText="No"
              >
                <Button
                  type={ButtonType?.Primary}
                  disabled={selectedNodeActionId === null}
                >
                  <FormattedMessage id="done" />
                </Button>
              </Popconfirm>
            </Col>
          </Row>
        }
      >
        <Row>
          <Col className="w-full">
            <Select
              placeholder={intl?.formatMessage({ id: 'select_node' })}
              className="w-full"
              onChange={handleChange}
              showSearch
            >
              {newNodeOptions?.length > 0 &&
                newNodeOptions.map((item: any) => {
                  return (
                    <Option value={item?.id} key={item.id}>{`${item?.id}_${
                      item?.name
                    }${
                      item?.node_action !== null ? '_' + item.node_action : ''
                    }`}</Option>
                  );
                })}
            </Select>
          </Col>
        </Row>
      </Modal>
      <Modal
        title={intl?.formatMessage({ id: 'select_node_action' })}
        centered
        destroyOnClose
        className="subnode-wrapper"
        visible={showModalChangeAction}
        onCancel={closeModalChangeAction}
        width={700}
        footer={
          <Row className="flex justify-end">
            <Col>
              <Popconfirm
                className="popconfrim"
                title="Are you sure？"
                onConfirm={() => handleChangeActionDone(nodeSubQuestion.id)}
                okText="Yes"
                cancelText="No"
              >
                <Button
                  type={ButtonType?.Primary}
                  disabled={selectedNodeActionId === null}
                  className="text-sm link px-1.5"
                >
                  <FormattedMessage id="done" />
                </Button>
              </Popconfirm>
            </Col>
          </Row>
        }
      >
        <Row>
          <Col className="w-full">
            <Select
              onChange={handleChangeAction}
              placeholder={intl?.formatMessage({ id: 'select_node_action' })}
              className="w-full"
              showSearch
            >
              {nodeActions?.length > 0 &&
                nodeActions.map((item: any) => {
                  return (
                    <Option
                      key={item.id}
                      value={item?.id}
                    >{`${item?.id}_${item?.header}`}</Option>
                  );
                })}
            </Select>
          </Col>
        </Row>
      </Modal>
    </div>
  );
};

export default React.memo(SubNode);
