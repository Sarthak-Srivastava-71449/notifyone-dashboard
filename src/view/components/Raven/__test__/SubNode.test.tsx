import React from 'react';
import { withProvider } from '../../../../common/renderWithProvider';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, fireEvent, screen } from '@testing-library/react';
import SubNode from '../SubNode';
import {
  accessToken,
  inactiveNodes,
  newNodeOptions,
  nodeActions,
  SubNodeData,
} from './mock';
import * as redux from 'react-redux';

jest.mock('src/store/actions/ravenActionNodeEvents');
jest.mock('src/store/actions/ravenNodeEvents');
jest.mock('src/store/actions/ravenRootNodeEvents');

// Raven Sub-Node-Unit-Testing Cases

describe('Raven Sub-Node-Test Suite', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  afterEach(cleanup);
  test('Match Create Node Snapshot', async () => {
    jest
      .spyOn(redux, 'useSelector')
      .mockImplementationOnce(() => accessToken)
      .mockImplementationOnce(() => inactiveNodes)
      .mockImplementationOnce(() => newNodeOptions)
      .mockImplementationOnce(() => nodeActions);
    const { asFragment } = render(
      withProvider(<SubNode nodeSubQuestion={undefined} rootNode={false} />)
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Showcase Mock Data', async () => {
    jest
      .spyOn(redux, 'useSelector')
      .mockImplementationOnce(() => accessToken)
      .mockImplementationOnce(() => inactiveNodes)
      .mockImplementationOnce(() => newNodeOptions)
      .mockImplementationOnce(() => nodeActions);

    render(
      withProvider(<SubNode nodeSubQuestion={undefined} rootNode={false} />)
    );
  });

  // Raven Rank Node Test Cases

  test('should edit `Enter Valid rank number` input when filling inputbox', () => {
    const component = render(withProvider(<SubNode nodeSubQuestion />));
    const input: any = component.getByTestId('rank-number');
    fireEvent.change(input, {
      target: { value: SubNodeData },
    });
    expect(SubNodeData).toBe(SubNodeData);
    expect(screen.queryByTestId('rank-number')).toBeInTheDocument();
  });
});
