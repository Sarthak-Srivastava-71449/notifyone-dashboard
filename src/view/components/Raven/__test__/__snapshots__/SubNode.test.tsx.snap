// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Raven Sub-Node-Test Suite Match Create Node Snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="ant-collapse ant-collapse-icon-position-start"
      role="tablist"
    >
      <div
        class="ant-collapse-item ant-collapse-item-disabled"
      >
        <div
          aria-disabled="true"
          aria-expanded="false"
          class="ant-collapse-header"
          role="tab"
          tabindex="-1"
        >
          <div
            class="ant-collapse-expand-icon"
          >
            <span
              aria-label="right"
              class="anticon anticon-right ant-collapse-arrow"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="right"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                />
              </svg>
            </span>
          </div>
          <span
            class="ant-collapse-header-text"
          >
            undefined undefined
          </span>
          <div
            class="ant-collapse-extra"
          >
            <div
              class="flex link items-center"
            >
              <div
                class="inner-item"
              >
                <div
                  class="px-1 font-semibold text-sm"
                >
                  <a
                    class="condition-btn px-1 py-1 link"
                  >
                    F
                  </a>
                </div>
                <div
                  class="ml-3"
                >
                  <input
                    class="ant-input custom-input input-number focus:ring-blue-500 focus:outline-none appearance-none w-full text-sm leading-6  rounded-md py-1 px-1.5  shadow-sm"
                    data-testid="rank-number"
                    placeholder="Rank"
                    type="number"
                    value=""
                  />
                </div>
                <div
                  class="ml-3"
                >
                  <a
                    class="link ml-1.5 px-1"
                  >
                    Change Action
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
